master/security/kubernetes-default-deny.md

```diff
@@ -0,0 +1,112 @@
+
+### Features
+
+This how-to guide uses the following Calico features:
+- **NetworkPolicy** 
+- **GlobalNetworkPolicy**
+
+### Concepts
+
+#### Default deny/allow behavior
+
+Default allow means all traffic is allowed by default, unless otherwise specified. Default deny means all traffic is denied by default, unless explicitly allowed. **Kubernetes pods are default allow, unless network policy is defined to specify otherwise**.
+
+For compatibility with Kubernetes, **Calico network policy** enforcement follows the standard convention for Kubernetes pods:
+- If no network policies apply to a pod, then all traffic to/from that pod is allowed.
+- If one or more network policies apply to a pod containing ingress rules, then only the ingress traffic specifically allowed by those policies is allowed, and all other traffic to/from the pod is denied.
+- If one or more network policies apply to a pod containing egress rules, then only the egress traffic specifically allowed by those policies is allowed, and all other traffic to/from the pod is denied.
+
+For other endpoint types (VMs, host interfaces), **Calico global network policy is default deny**. Only traffic specifically allowed by network policy is allowed, even if no network policies apply to the endpoint.
+
+#### Secure resources everywhere
+
+We recommend creating default deny using **Calico global network policy**, which is non-namespaced. It applies to both workloads (VMs and containers) and hosts (computers that run the hypervisor for VMs, or container runtime for containers). This supports a conservative security stance towards protecting resources.
+
+#### Calico network policy: order matters
+
+When you create Calico network policies to allow traffic, you can assign an **order** number that indicates the order you want policy applied relative to other policies. To ensure that your Calico default deny policy remains in effect at all times, assign a significantly higher number than regular policies that allow traffic.
+
+### How to
+
+Although you can use any of the following policies to create default deny for Kubernetes pods, we recommend using the Calico policy, non-namespaced.
+
+- [Create default deny-all ingress and egress Calico policy, non-namespaced](#create-default-deny-all-ingress-and-egress-calico-policy-non-namespaced)
+- [Create default deny-all ingress and egress Calico policy, namespaced](#create-default-deny-all-ingress-and-egress-calico-policy-namespaced)
+- [Create default deny all ingress and egress Kubernetes policy, namespaced](#create-default-deny-all-ingress-and-egress-Kubernetes-policy-namespaced)
+
+#### Create default deny-all ingress and egress Calico policy, non-namespaced
+
+In the following example, we specify a default deny global network policy that applies to workload endpoint resources in all namespaces, and host endpoint resources.  The **order: 500** is set high to ensure that allow traffic policies do not supersede it. 
```

Comment:
I don't think this is actually correct - most policies will default to a higher order than this (and thus, be executed _after_ this policy.

I think in order to implement a default-deny (i.e., and "implicit" deny) we would use a policy like this:

```
apiVersion: projectcalico.org/v3
kind: GlobalNetworkPolicy
metadata:
  name: default-deny
spec:
  selector: all()
  types:
  - Ingress
  - Egress
```

Reply:
The lack of rules means no traffic will be allowed by the policy. 

This is actually symmetric with how the k8s example works too.

Reply:
Thanks for the details. Fixed.